runtime/src/common/key_format.rs

```diff
@@ -74,6 +76,142 @@ pub trait KeyFormat {
     }
 }
 
+/// Part of the KeyFormat to be used with key-value backends for constructing keys.
+pub trait KeyFormatAtom {
+    fn size() -> usize;
+
+    fn encode_atom(self) -> Vec<u8>;
+
+    fn decode_atom(data: &[u8]) -> Self
+    where
+        Self: Sized;
+}
+
+impl KeyFormatAtom for u64 {
+    fn size() -> usize {
+        8
+    }
+
+    fn encode_atom(self) -> Vec<u8> {
+        self.to_be_bytes().to_vec()
+    }
+
+    fn decode_atom(data: &[u8]) -> Self
+    where
+        Self: Sized,
+    {
+        u64::from_be_bytes(data.try_into().expect("key_format: malformed u64 input"))
+    }
+}
+
+impl KeyFormatAtom for () {
+    fn size() -> usize {
+        0
+    }
+
+    fn encode_atom(self) -> Vec<u8> {
+        Vec::new()
+    }
+
+    fn decode_atom(_: &[u8]) -> () {
+        ()
+    }
+}
+
+impl<T1, T2> KeyFormatAtom for (T1, T2)
+where
+    T1: KeyFormatAtom,
+    T2: KeyFormatAtom,
+{
+    fn size() -> usize {
+        T1::size() + T2::size()
+    }
+
+    fn encode_atom(self) -> Vec<u8> {
+        let mut a = self.0.encode_atom();
+        let mut b = self.1.encode_atom();
+        a.append(&mut b);
+        a
+    }
+
+    fn decode_atom(data: &[u8]) -> (T1, T2) {
+        if data.len() < Self::size() {
+            panic!("key format atom: malformed input");
+        }
+
+        (
+            T1::decode_atom(&data[0..T1::size()]),
+            T2::decode_atom(&data[T2::size()..]),
+        )
+    }
+}
+
+// TODO: probably could generalize this for N sized tuples using macros.
```

Comment:
Would [`impl-trait-for-tuples`](https://crates.io/crates/impl-trait-for-tuples) help here?

Reply:
Yeah probably need to look into it. Staking module needed just up to 3 so i was fine implementing those manually for now :grin:

Reply:
fixed